File: tests/algorithms/binary_search/binary_search.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func binarySearch(arr []int, target int) int {
	left, right := 0, len(arr)-1
	
	for left <= right {
		mid := left + (right-left)/2
		
		if arr[mid] == target {
			return mid
		} else if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	
	return -1
}

func main() {
	size := 1000000
	arr := make([]int, size)
	for i := range arr {
		arr[i] = i
	}
	
	numSearches := 1000
	targets := make([]int, numSearches)
	for i := range targets {
		targets[i] = rand.Intn(size)
	}
	
	fmt.Printf("Performing %d binary searches on array of size %d...\n", numSearches, size)
	start := time.Now()
	
	foundCount := 0
	for _, target := range targets {
		if result := binarySearch(arr, target); result != -1 {
			foundCount++
		}
	}
	
	duration := time.Since(start)
	
	fmt.Printf("Result: Found %d/%d targets\n", foundCount, numSearches)
	fmt.Printf("Execution time: %.6f seconds\n", duration.Seconds())
}

File: tests/algorithms/binary_search/binary_search.rs
use std::time::Instant;
use rand::prelude::*;

fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    let mut left = 0;
    let mut right = arr.len();
    
    while left < right {
        let mid = left + (right - left) / 2;
        
        match arr[mid].cmp(&target) {
            std::cmp::Ordering::Equal => return Some(mid),
            std::cmp::Ordering::Less => left = mid + 1,
            std::cmp::Ordering::Greater => right = mid,
        }
    }
    
    None
}

fn main() {
    let size = 1_000_000;
    let arr: Vec<i32> = (0..size).collect();
    let num_searches = 1000;
    
    // Generate random targets
    let mut rng = thread_rng();
    let targets: Vec<i32> = (0..num_searches)
        .map(|_| rng.gen_range(0..size))
        .collect();
    
    println!("Performing {} binary searches on array of size {}...", num_searches, size);
    let start = Instant::now();
    
    let found_count = targets
        .iter()
        .filter(|&&target| binary_search(&arr, target).is_some())
        .count();
    
    let duration = start.elapsed();
    
    println!("Result: Found {}/{} targets", found_count, num_searches);
    println!("Execution time: {:.6} seconds", duration.as_secs_f64());
}